File: php/130-surrounded-regions.php
<?php
class Solution {
    private $row;
    private $col;
    private $direction = [
        [1, 0],
        [0, -1],
        [-1, 0],
        [0, 1]
    ];
    private $visited = [];
    /**
     * @param String[][] $board
     * @return NULL
     */
    function solve(&$board) {
        $this->row = count($board);
        $this->col = count($board[0]);
        $this->visited = array_fill(0, $this->row-1, array_fill(0, $this->col-1, false));
        for ($i = 0; $i < $this->row; $i++) {
            for ($j = 0; $j < $this->col; $j++) {
                if ($board[$i][$j] == 'O' && $this->onBorder($i, $j)) {
                    $this->dfs($board, $i, $j);
                }
            }
        }
        
        for ($i = 0; $i < $this->row; $i++) {
            for ($j = 0; $j < $this->col; $j++) {
                if (!$this->visited[$i][$j]) {
                    $board[$i][$j] = 'X';
                }
            }
        }
        return ;
    }
    function dfs(&$board, $x, $y) {
        if ($board[$x][$y] == 'X') {
            return true;
        }
        $this->visited[$x][$y] = true;
        for ($i = 0; $i < 4; $i++) {
            $nextX = $x + $this->direction[$i][0];
            $nextY = $y + $this->direction[$i][1];
            if ($this->inArea($nextX, $nextY) && !$this->visited[$nextX][$nextY]) {
                $this->dfs($board, $nextX, $nextY);
            }
        }
        return true;
    }
    function inArea($x, $y) {
        return $x >=0 && $x < $this->row && $y >= 0 && $y < $this->col;
    }
    function onBorder($x, $y) {
        if ($x == 0 || $x == $this->row-1 || $y == 0 || $y == $this->col-1) {
            return true;
        }
        return false;
    }
}

File: go/130-surrounded-regions.go
package main

var direction [][]int = [][]int{
	{1, 0},
	{0, -1},
	{-1, 0},
	{0, 1},
}
var visited [][]bool
var row, col int

func solve(board [][]byte) {
	if len(board) == 0 {
		return
	}
	row = len(board)
	col = len(board[0])
	//init visited
	visited = make([][]bool, row)
	for i := 0; i < row; i++ {
		visited[i] = make([]bool, col)
	}
	//mark border 'O'
	for j := 0; j < row; j++ {
		for k := 0; k < col; k++ {
			if board[j][k] == 'O' && onBorder(j, k) {
				dfs(&board, j, k)
			}
		}
	}
	//visited potin set to 'X'
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if !visited[i][j] {
				board[i][j] = 'X'
			}
		}
	}
	return
}
func dfs(board *[][]byte, x, y int) {
	if (*board)[x][y] == 'X' {
		return
	}
	visited[x][y] = true
	for i := 0; i < 4; i++ {
		nextX := x + direction[i][0]
		nextY := y + direction[i][1]
		if inArea(nextX, nextY) && !visited[nextX][nextY] {
			dfs(board, nextX, nextY)
		}
	}
}
func inArea(x, y int) bool {
	if x < 0 || x >= row || y < 0 || y >= col {
		return false
	}
	return true
}
func onBorder(x, y int) bool {
	return x == 0 || x == row-1 || y == 0 || y == col-1
}
